python/lsst/daf/butler/registry/collections/_base.py

```diff
@@ -0,0 +1,485 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = []
+
+from abc import abstractmethod
+import astropy.time
+import itertools
+from typing import (
+    Any,
+    Iterator,
+    NamedTuple,
+    Optional,
+    TYPE_CHECKING,
+)
+
+import sqlalchemy
+
+from ...core import ddl
+from ...core.timespan import Timespan, TIMESPAN_FIELD_SPECS
+from .._collectionType import CollectionType
+from ..interfaces import (
+    ChainedCollectionRecord,
+    CollectionManager,
+    CollectionRecord,
+    MissingCollectionError,
+    RunRecord,
+)
+from ..wildcards import CollectionSearch
+
+if TYPE_CHECKING:
+    from .database import Database
+
+
+def _makeCollForeignKey(sourceColumnName: str, collectionIdName: str, **kwargs) -> ddl.ForeignKeySpec:
+    """Define foreign key specification that refers to collections table.
+
+    Parameters
+    ----------
+    sourceColumnName : `str`
+        Name of the column in the referring table.
+    collectionIdName : `str`
+        Name of the column in collections table that identifies it (PK).
+    **kwargs
+        Additional keyword arguments passed directly to `ddl.ForeignKeySpec`.
+
+    Returns
+    -------
+    spec : `ddl.ForeignKeySpec`
+
+    Note
+    ----
+    This method assumes fixed name ("collection") of a collections table.
+    There is also a general assumption that collection primary key consists
+    of a single column.
```

Comment:
"Note" -> "Notes" (and in many other places).

But I really appreciate the clear descriptions of assumptions you're putting in these docs.